File: plugins/DungeonForge/src/main/java/nl/riddernix/dungeonforge/party/DungeonParty.java
package nl.riddernix.dungeonforge.party;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.UUID;

/** A transient party with insertion order used for leadership transfer. */
public final class DungeonParty {

    private final UUID id;
    private UUID leader;
    private final LinkedHashSet<UUID> members = new LinkedHashSet<>();

    DungeonParty(UUID leader) {
        this.id = UUID.randomUUID();
        this.leader = leader;
        members.add(leader);
    }

    public UUID id() {
        return id;
    }

    public UUID leader() {
        return leader;
    }

    public boolean isLeader(UUID playerId) {
        return leader.equals(playerId);
    }

    public Set<UUID> members() {
        return Set.copyOf(members);
    }

    public int size() {
        return members.size();
    }

    void add(UUID playerId) {
        members.add(playerId);
    }

    /** @return the new leader, or {@code null} when the party became empty */
    UUID remove(UUID playerId) {
        members.remove(playerId);
        if (members.isEmpty()) {
            return null;
        }
        if (leader.equals(playerId)) {
            leader = members.iterator().next();
        }
        return leader;
    }
}

File: plugins/DungeonPlugin/src/main/kotlin/nl/riddernix/dungeonplugin/party/DungeonParty.kt
package nl.riddernix.dungeonplugin.party

import java.util.UUID

/** A transient party with insertion order used for leadership transfer. */
class DungeonParty internal constructor(leader: UUID) {

    val id: UUID = UUID.randomUUID()
    var leader: UUID = leader
        private set
    private val memberSet = LinkedHashSet<UUID>()

    init {
        memberSet.add(leader)
    }

    fun isLeader(playerId: UUID): Boolean = leader == playerId

    val members: Set<UUID>
        get() = memberSet.toSet()

    val size: Int
        get() = memberSet.size

    internal fun add(playerId: UUID) {
        memberSet.add(playerId)
    }

    /** @return the new leader, or `null` when the party became empty */
    internal fun remove(playerId: UUID): UUID? {
        memberSet.remove(playerId)
        if (memberSet.isEmpty()) {
            return null
        }
        if (leader == playerId) {
            leader = memberSet.first()
        }
        return leader
    }
}
